mmedit/models/backbones/sr_backbones/liif_net.py

```diff
@@ -0,0 +1,259 @@
+
+        if self.local_ensemble:
+            vx_lst = [-1, 1]
+            vy_lst = [-1, 1]
+            eps_shift = 1e-6
+        else:
+            vx_lst, vy_lst, eps_shift = [0], [0], 0
+
+        # field radius (global: [-1, 1])
+        rx = 2 / feat.shape[-2] / 2
+        ry = 2 / feat.shape[-1] / 2
+
+        feat_coord = make_coord(feat.shape[-2:], flatten=False) \
+            .permute(2, 0, 1) \
+            .unsqueeze(0).expand(feat.shape[0], 2, *feat.shape[-2:])
+        feat_coord = feat_coord.to(coord)
+
+        preds = []
+        areas = []
+        for vx in vx_lst:
+            for vy in vy_lst:
+                coord_ = coord.clone()
+                coord_[:, :, 0] += vx * rx + eps_shift
+                coord_[:, :, 1] += vy * ry + eps_shift
+                coord_.clamp_(-1 + 1e-6, 1 - 1e-6)
+                query_feat = F.grid_sample(
+                    feat, coord_.flip(-1).unsqueeze(1),
+                    mode='nearest', align_corners=False)[:, :, 0, :] \
+                    .permute(0, 2, 1)
+                query_coord = F.grid_sample(
+                    feat_coord, coord_.flip(-1).unsqueeze(1),
+                    mode='nearest', align_corners=False)[:, :, 0, :] \
+                    .permute(0, 2, 1)
+                rel_coord = coord - query_coord
+                rel_coord[:, :, 0] *= feat.shape[-2]
+                rel_coord[:, :, 1] *= feat.shape[-1]
+                mid_tensor = torch.cat([query_feat, rel_coord], dim=-1)
+
+                if self.cell_decode:
+                    rel_cell = cell.clone()
+                    rel_cell[:, :, 0] *= feat.shape[-2]
+                    rel_cell[:, :, 1] *= feat.shape[-1]
+                    mid_tensor = torch.cat([mid_tensor, rel_cell], dim=-1)
+
+                bs, q = coord.shape[:2]
+                pred = self.imnet(mid_tensor.view(bs * q, -1)).view(bs, q, -1)
+                preds.append(pred)
+
+                area = torch.abs(rel_coord[:, :, 0] * rel_coord[:, :, 1])
+                areas.append(area + 1e-9)
+
+        tot_area = torch.stack(areas).sum(dim=0)
+        if self.local_ensemble:
+            t = areas[0]
+            areas[0] = areas[3]
+            areas[3] = t
+            t = areas[1]
+            areas[1] = areas[2]
+            areas[2] = t
+        result = 0
+        for pred, area in zip(preds, areas):
+            result = result + pred * (area / tot_area).unsqueeze(-1)
+
+        return result
+
+    def batched_predict(self, x, coord, cell):
+        """Batched predict.
+
+        Args:
+            x (Tensor): Input tensor.
+            coord (Tensor): coord tensor.
+            cell (Tensor): cell tensor.
+
+        Returns:
+            pred (Tensor): output of model.
+        """
+        with torch.no_grad():
+            n = coord.shape[1]
+            ql = 0
+            preds = []
+            while ql < n:
+                qr = min(ql + self.eval_bsize, n)
```

Comment:
I think the definition of `qr` is unncessary. When `qr > n`, `coord[:, ql:qr, :]` is equal to `coord[:, ql:, :]`

Reply:
define `right` will be better.